File: js/entities/world/Asteroid.js
// js/entities/world/Asteroid.js
import { Entity } from '../Entity.js';
import { state } from '../../state.js';
import { DRAG, PLANET_SPEED } from '../../constants.js';
import { Vector2 } from '../../vector2.js';
import { Sprite, Texture } from 'pixi.js';

export class Asteroid extends Entity {
  constructor(x, y, radius) {
    super();
    this.pos = new Vector2(x, y);
    this.radius = radius;
    this.mass = radius * radius;
    let direction = new Vector2(Math.random() * 2 - 1, Math.random() * 2 - 1).normalize();
    this.vel = direction.multiply(PLANET_SPEED);
    this.angularSpeed = (Math.random() * 2 - 1) * 0.05;
    this.angle = Math.random() * Math.PI * 2;
    this.color = '#A85417'; // ~10% brighter again on top of the previous '#994C15'
    this.points = this.generatePoints();

    // --- Generate interior points once for stable facets ---
    this.interiorPoints = [];
    const numInterior = 2 + Math.floor(Math.random() * 2); // 2-3 points
    for (let i = 0; i < numInterior; i++) {
      this.interiorPoints.push(new Vector2(
        (Math.random() - 0.5) * this.radius * 1.2,
        (Math.random() - 0.5) * this.radius * 1.2
      ));
    }
  }

  generatePoints() {
    const numSides = 6 + Math.floor(Math.random() * 6); // 6-11 sides
    const points = [];
    const angleStep = 2 * Math.PI / numSides;
    for (let i = 0; i < numSides; i++) {
      const a = i * angleStep + (Math.random() - 0.5) * angleStep * 0.5;
      const r = this.radius * (0.7 + Math.random() * 0.6);
      points.push(new Vector2(Math.cos(a) * r, Math.sin(a) * r));
    }
    return points;
  }

  update() {
    // Math.pow rather than a plain multiply for drag: DRAG is an
    // exponential per-frame decay rate, so scaling it correctly by
    // state.timeScale means raising it to that power, not multiplying
    // it — otherwise the asteroid's velocity would decay at its normal,
    // un-slowed rate even while its visible movement is slowed down,
    // and it wouldn't actually resume at the same velocity once V.A.T.S.
    // ends (see Player.js's applyPullForce for the fuller version of
    // this same reasoning, applied to gravity/pull acceleration).
    this.vel = this.vel.multiply(Math.pow(DRAG, state.timeScale));
    this.pos.add(this.vel.clone().multiply(state.timeScale));
    this.angle += this.angularSpeed * state.timeScale;

    // Bounce off walls
    if (this.pos.x - this.radius < 0) { this.pos.x = this.radius; this.vel.x = -this.vel.x; }
    if (this.pos.x + this.radius > state.sceneWidth) { this.pos.x = state.sceneWidth - this.radius; this.vel.x = -this.vel.x; }
    if (this.pos.y - this.radius < 0) { this.pos.y = this.radius; this.vel.y = -this.vel.y; }
    if (this.pos.y + this.radius > state.sceneHeight) { this.pos.y = state.sceneHeight - this.radius; this.vel.y = -this.vel.y; }
  }

  draw() {
    const ctx = state.ctx;
    ctx.save();
    ctx.translate(this.pos.x, this.pos.y);
    ctx.rotate(this.angle);

    const lightDir = new Vector2(-0.7, -0.7).normalize();
    const baseR = parseInt(this.color.substr(1, 2), 16);
    const baseG = parseInt(this.color.substr(3, 2), 16);
    const baseB = parseInt(this.color.substr(5, 2), 16);

    // --- Draw triangles along perimeter edges + interior points ---
    for (let i = 0; i < this.points.length; i++) {
      const j = (i + 1) % this.points.length;
      const p1 = this.points[i].clone();
      const p2 = this.points[j].clone();

      // Connect each interior point to the edge
      this.interiorPoints.forEach(ip => {
        this.fillTriangle(ctx, p1, p2, ip, baseR, baseG, baseB, lightDir);
      });

      // Optional subdivision along the edge itself for extra micro-facets
      const mid = p1.clone().add(p2).multiply(0.5);
      this.fillTriangle(ctx, p1, mid, p2, baseR, baseG, baseB, lightDir);
    }

    // --- Outline ---
    ctx.beginPath();
    ctx.moveTo(this.points[0].x, this.points[0].y);
    for (let i = 1; i < this.points.length; i++) {
      ctx.lineTo(this.points[i].x, this.points[i].y);
    }
    ctx.closePath();
    ctx.strokeStyle = '#3A1C08';
    ctx.lineWidth = 2;
    ctx.stroke();

    // --- Subtle inner shadow ---
    const shadowGrad = ctx.createRadialGradient(0, 0, 0, 0, 0, this.radius);
    shadowGrad.addColorStop(0, 'rgba(0,0,0,0)');
    shadowGrad.addColorStop(0.7, 'rgba(0,0,0,0.15)');
    shadowGrad.addColorStop(1, 'rgba(0,0,0,0.3)');
    ctx.fillStyle = shadowGrad;

    ctx.beginPath();
    ctx.moveTo(this.points[0].x, this.points[0].y);
    for (let i = 1; i < this.points.length; i++) {
      ctx.lineTo(this.points[i].x, this.points[i].y);
    }
    ctx.closePath();
    ctx.fill();

    ctx.restore();
  }

  // --- Helper to fill a triangle with shading ---
  fillTriangle(ctx, v0, v1, v2, baseR, baseG, baseB, lightDir) {
    const edge = v2.subtract(v1);
    const perp = new Vector2(edge.y, -edge.x);
    const normal = perp.lengthSq() > 0 ? perp.normalize() : new Vector2(0, 1);
    const dot = lightDir.dot(normal);

    const edgeDistance = (v1.length() + v2.length()) / (2 * this.radius);
    let brightness = 0.3 + Math.max(0, dot) * 0.7;
    brightness = brightness * (1 - 0.4 * edgeDistance) + 0.2;

    const cr = Math.min(255, Math.max(0, Math.floor(baseR * brightness)));
    const cg = Math.min(255, Math.max(0, Math.floor(baseG * brightness)));
    const cb = Math.min(255, Math.max(0, Math.floor(baseB * brightness)));

    ctx.fillStyle = `rgb(${cr},${cg},${cb})`;
    ctx.beginPath();
    ctx.moveTo(v0.x, v0.y);
    ctx.lineTo(v1.x, v1.y);
    ctx.lineTo(v2.x, v2.y);
    ctx.closePath();
    ctx.fill();
  }

  // ----------------------------
  // PIXI RENDERING (asteroids — see js/render/PixiStage.js for the
  // full migration context, and Planetoid.js's own
  // createPixiSprites/updatePixiSprites for the same overall pattern
  // applied there first). draw()/fillTriangle() above are completely
  // untouched.
  // ----------------------------
  // Unlike Planetoid.js, there's no existing offscreen bake to wrap —
  // draw() redraws live every frame. But this.points/interiorPoints/
  // color are all generated once in the constructor and never change
  // after that, so there's nothing gained by redrawing live here
  // either; this bakes the identical visual ONCE into a fresh offscreen
  // canvas — reusing fillTriangle() directly, since it already takes
  // ctx as a parameter rather than assuming state.ctx — and wraps that
  // as a Pixi texture, positioned/rotated per frame via sprite
  // properties instead of redrawn every frame.
  createPixiSprite(asteroidLayer) {
    if (this.pixiSprite) return;

    // Baked at a resolution higher than "zoom=1" needs, same reasoning
    // and same resMultiplier source as Player.js's own
    // initScaledAssets — so zooming in doesn't upscale a low-res bake.
    const resMultiplier = Math.max(1, state.zoomMax || 2.5);
    // Half-extent covers the max possible vertex distance
    // (generatePoints' own radius * up-to-1.3 multiplier) plus a
    // couple of pixels of margin for the 2px outline stroke.
    const halfExtent = this.radius * 1.4;
    const size = halfExtent * 2;
    const bakeSize = Math.max(1, Math.round(size * resMultiplier));

    const canvas = document.createElement('canvas');
    canvas.width = bakeSize;
    canvas.height = bakeSize;
    const bctx = canvas.getContext('2d');
    bctx.scale(resMultiplier, resMultiplier);
    bctx.translate(halfExtent, halfExtent); // this.points' own local origin (0,0) lands at canvas center

    const lightDir = new Vector2(-0.7, -0.7).normalize();
    const baseR = parseInt(this.color.substr(1, 2), 16);
    const baseG = parseInt(this.color.substr(3, 2), 16);
    const baseB = parseInt(this.color.substr(5, 2), 16);

    for (let i = 0; i < this.points.length; i++) {
      const j = (i + 1) % this.points.length;
      const p1 = this.points[i].clone();
      const p2 = this.points[j].clone();
      this.interiorPoints.forEach(ip => {
        this.fillTriangle(bctx, p1, p2, ip, baseR, baseG, baseB, lightDir);
      });
      const mid = p1.clone().add(p2).multiply(0.5);
      this.fillTriangle(bctx, p1, mid, p2, baseR, baseG, baseB, lightDir);
    }

    bctx.beginPath();
    bctx.moveTo(this.points[0].x, this.points[0].y);
    for (let i = 1; i < this.points.length; i++) bctx.lineTo(this.points[i].x, this.points[i].y);
    bctx.closePath();
    bctx.strokeStyle = '#3A1C08';
    bctx.lineWidth = 2;
    bctx.stroke();

    const shadowGrad = bctx.createRadialGradient(0, 0, 0, 0, 0, this.radius);
    shadowGrad.addColorStop(0, 'rgba(0,0,0,0)');
    shadowGrad.addColorStop(0.7, 'rgba(0,0,0,0.15)');
    shadowGrad.addColorStop(1, 'rgba(0,0,0,0.3)');
    bctx.fillStyle = shadowGrad;
    bctx.beginPath();
    bctx.moveTo(this.points[0].x, this.points[0].y);
    for (let i = 1; i < this.points.length; i++) bctx.lineTo(this.points[i].x, this.points[i].y);
    bctx.closePath();
    bctx.fill();

    const sprite = new Sprite(Texture.from(canvas));
    sprite.anchor.set(0.5, 0.5); // canvas center = this.points' own local origin, same point this.pos represents
    sprite.width = size;
    sprite.height = size;
    asteroidLayer.addChild(sprite);
    this.pixiSprite = sprite;
  }

  // Called every frame in place of draw() once migrated.
  updatePixiSprite(asteroidLayer) {
    if (!this.pixiSprite) {
      this.createPixiSprite(asteroidLayer);
      if (!this.pixiSprite) return;
    }
    this.pixiSprite.position.set(this.pos.x, this.pos.y);
    this.pixiSprite.rotation = this.angle;
  }
}

File: lua/entities/Asteroid.lua
-- lua/entities/Asteroid.lua
--
-- Port of js/entities/world/Asteroid.js
-- Irregular faceted asteroid with simple directional shading.

local state = require("lua.state")
local Vector2 = require("lua.vector2")
local constants = require("lua.constants")

local Asteroid = {}
Asteroid.__index = Asteroid

local function randomRange(a, b)
  return a + math.random() * (b - a)
end

function Asteroid.new(x, y, radius)
  local self = setmetatable({}, Asteroid)

  self.pos = Vector2.new(x, y)
  self.radius = radius
  self.mass = radius * radius

  local dir = Vector2.new(math.random() * 2 - 1, math.random() * 2 - 1):normalize()
  self.vel = dir:multiply(constants.PLANET_SPEED)

  self.angularSpeed = (math.random() * 2 - 1) * 0.05
  self.angle = math.random() * math.pi * 2

  -- Base brown color (roughly #A85417)
  self.color = { 0.66, 0.33, 0.09 }

  self.points = self:generatePoints()
  self.interiorPoints = {}

  local numInterior = 2 + math.floor(math.random() * 2) -- 2 or 3
  for i = 1, numInterior do
    table.insert(self.interiorPoints, Vector2.new(
      (math.random() - 0.5) * self.radius * 1.2,
      (math.random() - 0.5) * self.radius * 1.2
    ))
  end

  return self
end

function Asteroid:generatePoints()
  local numSides = 6 + math.floor(math.random() * 6) -- 6–11
  local points = {}
  local angleStep = (2 * math.pi) / numSides

  for i = 0, numSides - 1 do
    local a = i * angleStep + (math.random() - 0.5) * angleStep * 0.5
    local r = self.radius * (0.7 + math.random() * 0.6)
    table.insert(points, Vector2.new(math.cos(a) * r, math.sin(a) * r))
  end

  return points
end

function Asteroid:update()
  local ts = state.timeScale or 1

  -- Drag (same exponential reasoning as the JS version)
  local drag = constants.DRAG or 0.995
  self.vel = self.vel:multiply(drag ^ ts)

  self.pos:add(self.vel:clone():multiply(ts))
  self.angle = self.angle + self.angularSpeed * ts

  -- Bounce off world edges
  if self.pos.x - self.radius < 0 then
    self.pos.x = self.radius
    self.vel.x = -self.vel.x
  end
  if self.pos.x + self.radius > state.sceneWidth then
    self.pos.x = state.sceneWidth - self.radius
    self.vel.x = -self.vel.x
  end
  if self.pos.y - self.radius < 0 then
    self.pos.y = self.radius
    self.vel.y = -self.vel.y
  end
  if self.pos.y + self.radius > state.sceneHeight then
    self.pos.y = state.sceneHeight - self.radius
    self.vel.y = -self.vel.y
  end
end

-- Simple flat-shaded triangle
local function fillTriangle(x0, y0, x1, y1, x2, y2, baseR, baseG, baseB, lightDir, radius)
  -- Edge normal (approximate)
  local ex, ey = x2 - x1, y2 - y1
  local nx, ny = ey, -ex
  local len = math.sqrt(nx * nx + ny * ny)
  if len > 0 then
    nx, ny = nx / len, ny / len
  else
    nx, ny = 0, 1
  end

  local dot = lightDir.x * nx + lightDir.y * ny
  local edgeDist = (math.sqrt(x1*x1 + y1*y1) + math.sqrt(x2*x2 + y2*y2)) / (2 * radius)

  local brightness = 0.30 + math.max(0, dot) * 0.70
  brightness = brightness * (1 - 0.4 * edgeDist) + 0.20

  local r = math.min(1, math.max(0, baseR * brightness))
  local g = math.min(1, math.max(0, baseG * brightness))
  local b = math.min(1, math.max(0, baseB * brightness))

  love.graphics.setColor(r, g, b, 1)
  love.graphics.polygon("fill", x0, y0, x1, y1, x2, y2)
end

function Asteroid:draw()
  love.graphics.push()
  love.graphics.translate(self.pos.x, self.pos.y)
  love.graphics.rotate(self.angle)

  local lightDir = Vector2.new(-0.7, -0.7):normalize()
  local br, bg, bb = self.color[1], self.color[2], self.color[3]

  -- Faceted body
  for i = 1, #self.points do
    local j = (i % #self.points) + 1
    local p1 = self.points[i]
    local p2 = self.points[j]

    -- Connect to each interior point
    for _, ip in ipairs(self.interiorPoints) do
      fillTriangle(p1.x, p1.y, p2.x, p2.y, ip.x, ip.y, br, bg, bb, lightDir, self.radius)
    end

    -- Small mid-edge facet
    local mx = (p1.x + p2.x) * 0.5
    local my = (p1.y + p2.y) * 0.5
    fillTriangle(p1.x, p1.y, mx, my, p2.x, p2.y, br, bg, bb, lightDir, self.radius)
  end

  -- Dark outline
  love.graphics.setColor(0.23, 0.11, 0.03, 1)
  love.graphics.setLineWidth(2)
  local outline = {}
  for _, p in ipairs(self.points) do
    table.insert(outline, p.x)
    table.insert(outline, p.y)
  end
  love.graphics.polygon("line", unpack(outline))

  -- Subtle inner shadow (simple darkening toward the edge)
  love.graphics.setColor(0, 0, 0, 0.22)
  love.graphics.polygon("fill", unpack(outline))

  love.graphics.setColor(1, 1, 1, 1)
  love.graphics.setLineWidth(1)
  love.graphics.pop()
end

return Asteroid